File: extensions/vscode/src/commands.ts
/**
 * Every command the extension contributes, in one table.
 *
 * The ids here and the ids in package.json's `contributes.commands` are the same
 * list twice — a table makes that a diff anyone can read, instead of fifteen
 * `registerCommand` calls buried in the activation sequence.
 *
 * The three commands with a body of their own are the ones that ask a question:
 * the timings picker, the history picker and the export dialog. Everything else
 * is one call into the session.
 */
import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import * as vscode from 'vscode';

import { buildToolsImage, runDoctor } from './doctor';
import { Commit, delta, parseLog, parseTrend, sparkline, TrendEntry } from './history';
import { log } from './log';
import { gatesByDuration } from './parse';
import { probe } from './runner';
import { Session } from './session';

/** Commits `git log` is asked for when drawing the score history. */
const HISTORY_COMMITS = 40;

export function commandHandlers(session: Session): Record<string, () => unknown> {
  return {
    'gandalf.scanWorkspace': () => session.run('workspace'),
    'gandalf.scanCurrentFile': () => session.run('file'),
    'gandalf.showReport': () => session.openReport(false),
    'gandalf.showReportWithLlm': () => session.openReport(true),
    'gandalf.cancel': () => session.scheduler.cancel(),
    'gandalf.showLog': () => log().show(true),
    'gandalf.filterCurrentFile': () => session.findingsView.setScope('file'),
    'gandalf.filterProject': () => session.findingsView.setScope('project'),
    'gandalf.filterFindings': () => session.findingsView.pickFilters(),
    'gandalf.expandAll': () => session.findingsView.expandAll(),
    'gandalf.showTimings': () => showTimings(session),
    'gandalf.showHistory': () => showHistory(session),
    'gandalf.exportReport': () => exportReport(session),
    'gandalf.checkEnvironment': () => forFolder(session, runDoctor),
    'gandalf.buildToolsImage': () => forFolder(session, buildToolsImage),
  };
}

/** The commands that need somewhere to run do nothing when there is nowhere. */
async function forFolder(
  session: Session,
  action: (folder: vscode.WorkspaceFolder, s: ReturnType<Session['settingsFor']>) => Promise<void>,
): Promise<void> {
  const folder = session.primaryFolder();
  if (folder) await action(folder, session.settingsFor(folder));
}

/**
 * Per-gate timings, slowest first — the answer to "why does a full scan take so
 * long". Selecting gates copies a ready-to-paste skip list, since trimming the
 * gate set is the only real lever on a slow scan.
 */
async function showTimings(session: Session): Promise<void> {
  const folder = session.primaryFolder();
  const snapshot = folder ? session.store.project(folder) : undefined;
  const lastRun = folder ? session.store.lastRun(folder) : undefined;
  const timed = gatesByDuration(snapshot?.payload);
  if (!snapshot || timed.length === 0) {
    void vscode.window.showInformationMessage(
      'Gandalf: no timings yet — run “Gandalf: Scan Workspace” first.',
    );
    return;
  }
  const summed = timed.reduce((n, g) => n + (g.duration ?? 0), 0);
  const wall = (lastRun?.durationMs ?? 0) / 1000;
  const chosen = await vscode.window.showQuickPick(
    timed.map((g) => ({
      label: g.name,
      description: `${(g.duration ?? 0).toFixed(1)}s`,
      detail: g.summary,
    })),
    {
      canPickMany: true,
      title: `Gate timings — ${wall.toFixed(1)}s wall clock, ${summed.toFixed(1)}s summed (gates run concurrently)`,
      placeHolder: 'Select gates to copy a .gandalf.toml skip list for editor scans',
    },
  );
  if (!chosen?.length) return;
  await vscode.env.clipboard.writeText(
    `[gandalf]\nskip = [${chosen.map((c) => `"${c.label}"`).join(', ')}]\n`,
  );
  void vscode.window.showInformationMessage(
    `Gandalf: copied a skip list for ${chosen.length} gate(s). Paste it into a .gandalf.toml and ` +
      'point `gandalf.configPath` at that file to use it for editor scans only.',
  );
}

/** The scores in `.gandalf-trend.jsonl`, or an empty history when there is none. */
async function readTrend(root: string): Promise<Map<string, TrendEntry>> {
  try {
    return parseTrend(await fs.promises.readFile(path.join(root, '.gandalf-trend.jsonl'), 'utf8'));
  } catch {
    // No log yet — the picker still lists commits, all unscored.
    return new Map();
  }
}

/** One quick pick item per commit, scored ones carrying their change. */
function historyItems(
  commits: Commit[],
  trend: Map<string, TrendEntry>,
  previous: Map<string, number>,
) {
  return commits.map((c) => {
    const entry = trend.get(c.short);
    const change = entry ? delta(entry.score, previous.get(c.short)) : '';
    return {
      label: entry ? `${entry.score}/100${change ? `  ${change}` : ''}` : '— not scanned',
      description: `${c.short}  ${c.subject}`,
      detail: c.date,
      commit: c.short,
    };
  });
}

/**
 * Score over time. `.gandalf-trend.jsonl` holds what CLI and CI runs recorded;
 * `git log` supplies the commits, including the ones nothing has scored yet,
 * because "we have never measured this" is part of the history too.
 */
async function showHistory(session: Session): Promise<void> {
  const folder = session.primaryFolder();
  if (!folder) return;
  const root = folder.uri.fsPath;

  const trend = await readTrend(root);
  const gitLog = await probe('git', ['log', `-${HISTORY_COMMITS}`, '--format=%h%x1f%s%x1f%cs'], root);
  const commits: Commit[] = gitLog.ok ? parseLog(gitLog.output) : [];
  if (commits.length === 0) {
    void vscode.window.showInformationMessage('Gandalf: no commits to show a history for.');
    return;
  }

  // Oldest first for the sparkline and the deltas; newest first in the list.
  const scored = [...commits].reverse().filter((c) => trend.has(c.short));
  const line = sparkline(scored.map((c) => trend.get(c.short)!.score));
  const previous = new Map<string, number>();
  scored.forEach((c, i) => {
    if (i > 0) previous.set(c.short, trend.get(scored[i - 1].short)!.score);
  });

  const chosen = await vscode.window.showQuickPick(historyItems(commits, trend, previous), {
    title: scored.length
      ? `Score history — ${line} over ${scored.length} scanned commit(s) of ${commits.length}`
      : `Score history — nothing scanned yet of ${commits.length} commit(s)`,
    placeHolder: 'Pick a commit to scan it and open its report',
  });
  if (!chosen) return;
  const ok = await session.run('commit', {
    commit: chosen.commit,
    reason: `commit ${chosen.commit}`,
  });
  if (ok && session.reportView.current) {
    await session.reportView.show(session.reportView.current, chosen.commit);
  }
}

/**
 * Save the scorecard somewhere the user chooses — gandalf writes it into the
 * extension's storage, which is no use for attaching to a pull request.
 */
async function exportReport(session: Session): Promise<void> {
  if (!session.reportView.current || !fs.existsSync(session.reportView.current)) {
    const ok = await session.run('workspace', { reason: 'export' });
    if (!ok || !session.reportView.current) return;
  }
  const folder = session.primaryFolder();
  const suggested = path.basename(session.reportView.current);
  const target = await vscode.window.showSaveDialog({
    defaultUri: vscode.Uri.file(path.join(folder?.uri.fsPath ?? os.homedir(), suggested)),
    filters: { 'HTML report': ['html'] },
    title: 'Export the Gandalf report',
  });
  if (!target) return;
  await fs.promises.copyFile(session.reportView.current, target.fsPath);
  const choice = await vscode.window.showInformationMessage(
    `Gandalf: report exported to ${path.basename(target.fsPath)}.`,
    'Open',
  );
  if (choice === 'Open') await vscode.env.openExternal(target);
}

File: extensions/nvim/lua/gandalf/commands.lua
-- What the :Gandalf* commands do.
--
-- Each one needs a completed run to say anything, so each goes through
-- `ensure_scanned`: if nothing has been scanned yet, scan first and continue
-- only if that worked.

local core = require('gandalf.core')
local scan = require('gandalf.scan')
local state = require('gandalf.state')
local ui = require('gandalf.ui')

local M = {}

--- Commits `git log` is asked for when drawing the score history.
local HISTORY_COMMITS = 40

local function ensure_scanned(cb)
  if state.snapshot() then
    return cb()
  end
  state.notify('scanning…')
  scan.run({
    kind = 'workspace',
    manual = true,
    reason = 'report',
    on_done = function(ok)
      if ok then
        cb()
      end
    end,
  }, scan.cancel)
end

function M.report()
  ensure_scanned(function()
    ui.float(ui.report_lines(state.snapshot()), {
      title = ' Gandalf scorecard ',
      filetype = 'gandalf-report',
    })
  end)
end

--- Explain the dependency under the cursor: what gandalf found against it, and
--- what it could not check.
function M.hover()
  local package = core.package_at_cursor(vim.api.nvim_get_current_line())
  if package == '' then
    return state.notify('no dependency on this line.')
  end
  ensure_scanned(function()
    local matched = core.findings_for_package(state.findings(), package)
    ui.float(ui.hover_lines(package, matched, state.snapshot()), {
      title = ' ' .. package .. ' ',
      filetype = 'gandalf-hover',
    })
  end)
end

--- Every finding, in the quickfix list.
function M.list()
  ensure_scanned(function()
    local findings = state.findings()
    if #findings == 0 then
      return state.notify('nothing found.')
    end
    ui.to_quickfix(findings, scan.root(), 'Gandalf')
    vim.cmd('copen')
  end)
end

--- How many findings sit at each level, and which levels have any.
local function levels_present(findings)
  local counts = {}
  for _, finding in ipairs(findings) do
    counts[finding.level] = (counts[finding.level] or 0) + 1
  end
  local present = {}
  for _, level in ipairs(core.LEVELS) do
    if (counts[level] or 0) > 0 then
      present[#present + 1] = level
    end
  end
  return present, counts
end

--- Pick a level; its findings go to the quickfix list.
function M.filter()
  ensure_scanned(function()
    local all = state.findings()
    local levels, counts = levels_present(all)
    if #levels == 0 then
      return state.notify('nothing found.')
    end
    vim.ui.select(levels, {
      prompt = 'Show findings at level',
      format_item = function(level)
        return string.format('%-9s %d', core.LEVEL_LABEL[level], counts[level])
      end,
    }, function(level)
      if not level then
        return
      end
      local rows = {}
      for _, finding in ipairs(all) do
        if finding.level == level then
          rows[#rows + 1] = finding
        end
      end
      ui.to_quickfix(rows, scan.root(), 'Gandalf: ' .. core.LEVEL_LABEL[level])
      vim.cmd('copen')
    end)
  end)
end

--- Where the time went. Selecting a gate copies a ready-to-paste skip list,
--- since trimming the gate set is the only real lever on a slow scan.
function M.timings()
  local snapshot = state.snapshot()
  if not snapshot then
    return state.notify('no timings yet — run :GandalfScanAll first.')
  end
  local timed = {}
  for _, gate in ipairs(snapshot.payload.gates or {}) do
    if type(gate.duration) == 'number' then
      timed[#timed + 1] = gate
    end
  end
  if #timed == 0 then
    return state.notify('this run recorded no gate timings.')
  end
  table.sort(timed, function(a, b)
    return a.duration > b.duration
  end)
  vim.ui.select(timed, {
    prompt = 'Gate timings (slowest first) — pick one to copy a skip list',
    format_item = function(gate)
      return string.format('%-24s %6.1fs  %s', gate.name, gate.duration, gate.summary or '')
    end,
  }, function(gate)
    if not gate then
      return
    end
    local snippet = string.format('[gandalf]\nskip = ["%s"]\n', gate.name)
    vim.fn.setreg('+', snippet)
    vim.fn.setreg('"', snippet)
    state.notify(
      ('copied a skip list for %s. Paste it into a .gandalf.toml and point '):format(gate.name)
        .. 'config_path at that file to use it for editor scans only.'
    )
  end)
end

--- The scores in .gandalf-trend.jsonl, or an empty history when there is none.
local function read_trend(root)
  local log_path = vim.fs.joinpath(root, '.gandalf-trend.jsonl')
  if not vim.uv.fs_stat(log_path) then
    return {}
  end
  return core.parse_trend(table.concat(vim.fn.readfile(log_path), '\n'), vim.json.decode)
end

--- The scored commits oldest first, and the score each one moved from. Both are
--- for the sparkline and the deltas, which read forwards through history.
local function scored_series(commits, trend)
  local shorts, scores = {}, {}
  for i = #commits, 1, -1 do
    local entry = trend[commits[i].short]
    if entry then
      shorts[#shorts + 1] = commits[i].short
      scores[#scores + 1] = entry.score
    end
  end
  local previous = {}
  for i, short in ipairs(shorts) do
    if i > 1 then
      previous[short] = scores[i - 1]
    end
  end
  return scores, previous
end

local function history_prompt(scores, commits)
  if #scores > 0 then
    return string.format(
      'Score history — %s over %d of %d commit(s)',
      core.sparkline(scores),
      #scores,
      #commits
    )
  end
  return string.format('Score history — nothing scanned yet of %d commit(s)', #commits)
end

local function scan_commit(short)
  state.notify('scanning ' .. short .. '…')
  scan.run({
    kind = 'commit',
    commit = short,
    manual = true,
    reason = 'commit ' .. short,
    on_done = function(ok)
      if ok then
        M.report()
      end
    end,
  }, scan.cancel)
end

--- Score over time, from the log gandalf's own runs keep. `git log` supplies
--- the commits, including the ones nothing has scored yet, because "we have
--- never measured this" is part of the history too.
function M.history()
  local root = scan.root()
  local trend = read_trend(root)

  vim.system(
    { 'git', 'log', '-' .. HISTORY_COMMITS, '--format=%h%x1f%s%x1f%cs' },
    { text = true, cwd = root },
    function(out)
      vim.schedule(function()
        local commits = out.code == 0 and core.parse_log(out.stdout or '') or {}
        if #commits == 0 then
          return state.notify('no commits to show a history for.')
        end
        local scores, previous = scored_series(commits, trend)
        vim.ui.select(commits, {
          prompt = history_prompt(scores, commits),
          format_item = function(commit)
            local entry = trend[commit.short]
            local label = entry
                and string.format('%d/100 %-4s', entry.score, core.delta(entry.score, previous[commit.short]))
              or '— not scanned'
            return string.format('%-16s %s  %s', label, commit.short, commit.subject)
          end,
        }, function(commit)
          if commit then
            scan_commit(commit.short)
          end
        end)
      end)
    end
  )
end

function M.show_log()
  local lines = state.log_text()
  ui.float(#lines > 0 and lines or { 'nothing logged yet' }, {
    title = ' Gandalf log ',
    filetype = 'log',
  })
end

return M
